subprojects/gdk_rust/src/lib.rs

```diff
@@ -0,0 +1,546 @@
+                last_xr: None,
+            };
+            Ok(gdk_session)
+        }
+        _ => Err(json!("server_type invalid")),
+    }
+}
+
+fn fetch_cached_exchange_rates(sess: &mut GdkSession) -> Option<Vec<Ticker>> {
+    if sess.last_xr.is_some()
+        && (SystemTime::now() < (sess.last_xr_fetch + Duration::from_secs(60)))
+    {
+        debug!("hit exchange rate cache");
+    } else {
+        info!("missed exchange rate cache");
+        let rates = fetch_exchange_rates();
+        // still record time even if we get no results
+        sess.last_xr_fetch = SystemTime::now();
+        if !rates.is_empty() {
+            // only set last_xr if we got new non-empty rates
+            sess.last_xr = Some(rates);
+        }
+    }
+
+    sess.last_xr.clone()
+}
+
+#[no_mangle]
+pub extern "C" fn GDKRUST_call_session(
+    sess: *mut GdkSession,
+    method: *const c_char,
+    input: *const GDKRUST_json,
+    output: *mut *const GDKRUST_json,
+) -> i32 {
+    let method = read_str(method);
+    let input = &safe_ref!(input).0;
+
+    let sess = safe_mut_ref!(sess);
+
+    if method == "exchange_rates" {
+        let rates = fetch_cached_exchange_rates(sess).unwrap_or_default();
+        return json_res!(output, tickers_to_json(rates), GA_OK);
+    }
+
+    // Redact inputs containing private data
+    let methods_to_redact = vec!["login", "register_user", "set_pin", "create_subaccount"];
+    let input_str = format!("{:?}", &input);
+    let input_redacted = if methods_to_redact.contains(&method.as_str())
+        || input_str.contains("mnemonic")
+        || input_str.contains("xprv")
+    {
+        "redacted".to_string()
+    } else {
+        input_str
+    };
+
+    info!("GDKRUST_call_session handle_call {} input {:?}", method, input_redacted);
+    let res = match sess.backend {
+        GdkBackend::Electrum(ref mut s) => handle_call(s, &method, &input),
+        // GdkSession::Rpc(ref s) => handle_call(s, method),
+    };
+
+    let mut res_string = format!("{:?}", res);
+    res_string.truncate(200);
+    info!("GDKRUST_call_session {} {:?}", method, res_string);
+
+    match res {
+        Ok(ref val) => json_res!(output, val, GA_OK),
+
+        Err(ref e) => {
+            let code = e.to_gdk_code();
+            let desc = e.gdk_display();
+
+            info!("rust error {}: {}", code, desc);
+            json_res!(output, json!({ "error": code, "message": desc }), GA_OK)
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn GDKRUST_set_notification_handler(
+    sess: *mut GdkSession,
+    handler: extern "C" fn(*const libc::c_void, *const GDKRUST_json),
+    self_context: *const libc::c_void,
+) -> i32 {
+    let sess = safe_mut_ref!(sess);
+    let backend = &mut sess.backend;
+
+    match backend {
+        GdkBackend::Electrum(ref mut s) => s.notify = NativeNotif(Some((handler, self_context))),
+    };
+
+    info!("set notification handler");
+
+    GA_OK
+}
+
+fn fetch_exchange_rates() -> Vec<Ticker> {
```

Comment:
Would be nice if this supported requesting via Tor out of the box. Green already comes bundled with Tor, so this should be relatively straightforward, right?

I checked now and it appears that the bitfinex api endpoint accepts requests coming from Tor (or at least, doesn't ban them entirely).

There's also an exchange rate api operated by wasabi as a hidden service that could be considered: http://wasabiukrxmkdgve5kynjztuovbg43uxcbcxn6y2okcrsg7gb6jdmbad.onion/api/v3/btc/Offchain/exchange-rates

Reply:
Not sure it would be straightforward at least for me, but I didn't follow tor lib integration in gdk cpp, also it would need some new bindings to start/stop tor service or other architecture.

Reply:
I imagined that it would already be running if the user enabled tor support, so no need to start/stop, just to figure out if its running and on what port. Maybe @greenaddress can tell us more about this?